digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/RoadNameService.scala

```diff
@@ -1,29 +1,102 @@
 package fi.liikennevirasto.viite
 
 import fi.liikennevirasto.digiroad2.oracle.OracleDatabase
-import fi.liikennevirasto.viite.dao.{RoadName, RoadNameDAO}
+import fi.liikennevirasto.digiroad2.user.User
+import fi.liikennevirasto.viite.dao.{ProjectLinkNameDAO, RoadName, RoadNameDAO}
 import org.joda.time.DateTime
 import org.slf4j.LoggerFactory
 
 import scala.util.control.NonFatal
 
+case class RoadNameRows(roadId: Long, editions: Seq[RoadNameEditions])
+
+case class RoadNameEditions(editedField: String, value: String)
+
 class RoadNameService() {
 
   def withDynTransaction[T](f: => T): T = OracleDatabase.withDynTransaction(f)
 
+  def withDynSession[T](f: => T): T = OracleDatabase.withDynSession(f)
+
   private val logger = LoggerFactory.getLogger(getClass)
 
   private def queryRoadNamesAndNumbers(oRoadNumber: Option[Long], oRoadName: Option[String],
                                        oStartDate: Option[DateTime] = None, oEndDate: Option[DateTime] = None): Seq[RoadName] = {
     RoadNameDAO.getRoadNamesByRoadNameAndRoadNumber(oRoadNumber, oRoadName, None, None, oStartDate, oEndDate)
   }
 
+  import org.joda.time.DateTime
+  import org.joda.time.format.DateTimeFormat
+
+  val formatter = DateTimeFormat.forPattern("dd.MM.yyyy")
+
+  private def singleEndDateExpiration(nameRows: Seq[RoadNameRows]) = {
+    val field = decodeFields(nameRows.head.editions)
+    val invalid = nameRows.size == 1 && nameRows.head.roadId != NewRoadName && field("endDate").nonEmpty
+    (invalid, nameRows.head.roadId)
+  }
+
+  private def singleNameExpiration(nameRows: Seq[RoadNameRows]) = {
+    val field = decodeFields(nameRows.head.editions)
+    val toBeExpired = nameRows.size == 1 && nameRows.head.roadId != NewRoadName && field.size == 1 && field.head._1 == "roadName"
+    (toBeExpired, field.head._2)
+  }
+
   def getRoadAddresses(oRoadNumber: Option[String], oRoadName: Option[String], oStartDate: Option[DateTime], oEndDate: Option[DateTime]): Either[String, Seq[RoadName]] = {
     withDynTransaction {
       getRoadAddressesInTX(oRoadNumber, oRoadName, oStartDate, oEndDate)
     }
   }
 
+  def addOrUpdateRoadNames(roadNames: Seq[RoadNameRows], user: User): Option[String] = {
+    withDynTransaction {
+      addOrUpdateRoadNamesInTX(roadNames, user)
+    }
+  }
+
+  def addOrUpdateRoadNamesInTX(roadNames: Seq[RoadNameRows], user: User): Option[String] = {
```

Comment:
Each time we are editing we should just expire and create a new entry with the new road name values.